test/quantization/test_workflow_module.py

```diff
@@ -862,22 +862,36 @@ def test_backward_per_tensor(self, device, X):
         Y_prime.backward(dout)
         np.testing.assert_allclose(dX.cpu(), X.grad.cpu().detach().numpy(), rtol=tolerance, atol=tolerance)
 
-    def _test_forward_per_tensor_cachemask_impl(self, device):
-        for torch_type in (torch.qint8, torch.quint8):
-            Xs = (torch.randn(4, 8, device=device), torch.randn(4, 16, device=device)[:, ::2])
-            # pick the scale + zp so that some values get clipped
-            for X in Xs:
-                obs = torch.quantization.MinMaxObserver(torch_type)
-                obs(X * 0.75)
-                scale, zero_point = obs.calculate_qparams()
-                scale, zero_point = float(scale), int(zero_point)
-                quant_min, quant_max = obs._calculate_qmin_qmax()
+    def test_forward_backward_per_tensor_cachemask_with_amp(self):
+        self.assertTrue(torch.cuda.is_available())
+        net = nn.Sequential(nn.Conv2d(1, 1, 3))
+        net.qconfig = torch.quantization.get_default_qat_qconfig('fbgemm')
+        net_prep = torch.quantization.prepare_qat(net)
+
+        with torch.cuda.amp.autocast():
+            x = torch.randn(4, 1, 5, 5)
+            out = net_prep(x).sum()
+            out.backward()
+            self.assertTrue(net_prep[0].weight.grad is not None)
 
-                Y_test = torch.fake_quantize_per_tensor_affine(
-                    X, scale, zero_point, quant_min, quant_max)
-                Y_ref = _fake_quantize_per_tensor_affine_reference(
-                    X.cpu(), scale, zero_point, quant_min, quant_max).to(device)
-                self.assertTrue(torch.allclose(Y_test, Y_ref, rtol=tolerance, atol=tolerance))
+    def _test_forward_per_tensor_cachemask_impl(self, device):
+        for float_type in (torch.float32, torch.float16, torch.float64):
+            for torch_type in (torch.qint8, torch.quint8):
+                Xs = (torch.randn(4, 8, device=device), torch.randn(4, 16, device=device)[:, ::2])
+                # pick the scale + zp so that some values get clipped
+                for X in Xs:
+                    X = X.to(float_type)
+                    obs = torch.quantization.MinMaxObserver(torch_type)
+                    obs(X * 0.75)
+                    scale, zero_point = obs.calculate_qparams()
```

Comment:
Since you are interested only in the numerics of the fake-quant operator, you do not need an observer.  You can directly test the operator with the values of scale and zero-point that you need. Ideally, you could sweep over a set of values for scale and zero-point. i.e something like
```
for scale, zero_point in itertools.product(scale_values, zero_point_values):
```

Reply:
That is from previous diffs where the observer is used to pick a  quant_min and quant_max such that it is guaranteed to clip the matrix. Its not used in the actual quantization test. My only change to this section is to add a for loop over torch type and apply it to X.